src/main/scala/io/picnicml/doddlemodel/typeclasses/Clusterer.scala

```diff
@@ -0,0 +1,35 @@
+package io.picnicml.doddlemodel.typeclasses
+
+import io.picnicml.doddlemodel.data.Features
+
+trait Clusterer[A] extends Estimator[A] {
+
+  def fit(model: A, x: Features): A = {
+    require(!isFitted(model), "Called fit on a model that is already fitted")
+    fitSafe(copy(model), x)
+  }
+
+  def fitPredict(model: A, x: Features): Array[Int] = {
+    require(!isFitted(model), "Called fit on a model that is already fitted")
+    fitPredictSafe(copy(model), x)
+  }
+
+  def label(model: A): Array[Int] = {
+    require(isFitted(model), "Request label on a model that is not fitted yet")
+    labelSafe(model)
+  }
+
+  /** A function that creates an identical clusterer. */
+  protected def copy(model: A): A
+  protected def copy(model: A, label: Array[Int]): A
```

Comment:
The `copy(model: A, label: Array[Int])` is necessary in the current structure.
Otherwise the result `label` will not be kept in the next copy because it is just a local variable.